Post By Email: add front-end settings for Simple sites

chad1008 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/7210

## Proposed Changes

- moves Post by Email module and settings into their own file
- implements a Simple Site version of those settings utilizing the endpoints being introduced for 7211-gh-Automattic/dotcom-forge

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Post by email is currently only available in Calypso for Atomic sites, but [it appears this may not have been an intentional choice](6941-gh-Automattic/dotcom-forge). As work is done on Post by Voice settings, it makes sense to update the nearby Post by Email at the same time.

## Notes

- This PR took a lot of inspiration from @nightnei's Post by Voice work in #90489. Before I decided to split these settings off into a separate file I actually branched off of that PR so I could more easily keep any changes in sync, so we actually have at least one commit in common at the moment. I'll rebase off of that PR periodically to keep on the same page.
- Currently in a fairly raw state, it needs a lot of cleanup. I liberally applied the props I needed to silence TypeScript, and will now be going through and trimming them down to what's actually needed
- The settings are (naturally) not functional as of yet, that'll wait until the new endpoints are in place
- For the setting toggle, I'm currently rendering either a `JetpackModuleToggle` as the setting did previously, or just a `ToggleControl` from @wordpress/components. I'd like to look into whether or not we can use the `ToggleControl` in both cases and just modify its handlers for Atomic/Jetpack vs Simple sites.

## Testing Instructions

- On your sandbox, apply the patch from 7211-gh-Automattic/dotcom-forge
- Point public-api.wordpress.com to your sandbox
- Running Calypso locally off of this branch, visit **Settings > Writing** for a simple site, and make sure you have the **Classic View** selected (Calypso, not wp-admin)
- Confirm that the Post by Email settings appear
- In a separate tab, open https://dashboard.wordpress.com/
- Find your simple test site in the dashboard and note the state and/or email address of the current post by email settings
- Back in Calypso, test the Post by Email toggle, refreshing and checking dashboard.wordpress.com to confirm the settings from Calypso are carrying over.
- Back in Calypso, test the Regenerate button. Confirm that a new address is generated, and that the new address if reflected when refreshing dashboard.wordpress.com
- Switch over to **Settings > Writing** for a WoA site
- Repeat the same tests, this time comparing them to `wp-admin/admin.php?page=jetpack#/writing`, making sure the Calypso settings still properly update the Jetpack settings (we didn't make changes here, but the code was moved and refactored so we're making sure nothing got broken in the process)

nightnei committed:
feat: use real endpoints

nightnei committed:
feat: do not render PbV if uploading file is disallowed

nightnei committed:
chore: revert userCan changes

chad1008 committed:
inital pass moving pbe to separate file

chad1008 committed:
inital framework for simple calypso settings

chad1008 committed:
clean up props

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,160 @@
+
+export const PostByEmailSetting = connect(
+	( state: IAppState ) => {
+		const selectedSiteId = getSelectedSiteId( state ) || 0;
+		const siteIsJetpack = isJetpackSite( state, selectedSiteId );
+		const isAtomic = isSiteAutomatedTransfer( state, selectedSiteId );
+		const regeneratingPostByEmail = isRegeneratingJetpackPostByEmail( state, selectedSiteId );
+		const active = !! isJetpackModuleActive( state, selectedSiteId, moduleSlug );
+		const siteInDevMode = isJetpackSiteInDevelopmentMode( state, selectedSiteId );
+		const moduleUnavailableInDevMode = isJetpackModuleUnavailableInDevelopmentMode(
+			state,
+			selectedSiteId,
+			moduleSlug
+		);
+
+		return {
+			selectedSiteId,
+			siteIsJetpack,
+			isAtomic,
+			regeneratingPostByEmail,
+			active,
+			moduleUnavailable: siteInDevMode && moduleUnavailableInDevMode,
+		};
```
Same as here https://github.com/Automattic/wp-calypso/pull/90489#discussion_r1603539659, we should use `useSelector` from `calypso/state` instead of `connect`s `mapStateToProps`.

nightnei replied:
@fredrikekelund In my case it was just 1 property and I thought that it was the reason to get rid of `connect`. But looks like the reason is something else.
Is there a recommendation from A8C to avoid using `connect` or maybe some other reason? Since, personally, for me, `connect` is quite good tool.

And with the help of `useSelector` it does not look terse when we copy variable names 3 times, doing simple things.
https://github.com/Automattic/wp-calypso/pull/90819/files#diff-1a2d2a711ac2389148e7764b72177d565e3e6a88fea709f208906a031a5a64c5R36-R64


@chad1008 Let's at least avoid redundant variable renamings:
```
const {
		selectedSiteId,
        ...
	} = useSelector( ( state ) => {
		const selectedSiteId = getSelectedSiteId( state ) || 0;
        ....

		return {
			selectedSiteId,
			....
		};
	} );
```

Instead of:
```
const {
		siteId: selectedSiteId,
        ...
	} = useSelector( ( state ) => {
		const siteId = getSelectedSiteId( state ) || 0;
        ....

		return {
			siteId,
			....
		};
	} );
```

chad1008 replied:
Oh yes. I have a lot of variable renaming to do. Stuff like this, but I'm also going through and making sure it's clear which flow (Simple vs Jetpack) individual variables are a part of.

fredrikekelund replied:
I left a comment about this here https://github.com/Automattic/wp-calypso/pull/90819#discussion_r1611276938

fredrikekelund replied:
> Is there a recommendation from A8C to avoid using connect or maybe some other reason? Since, personally, for me, connect is quite good tool.

`connect` isn't deprecated, it's just that hooks are recommended as the default (see [docs](https://react-redux.js.org/api/connect)).

This post (even if it's old) has some Calypso-specific background p4TIVU-9k1-p2
In short, `connect` is a viable alternative for more complex cases. For this particular PR, I think we should stick with `useSelector`.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,160 @@
+			moduleUnavailable: siteInDevMode && moduleUnavailableInDevMode,
+		};
+	},
+	{
+		activateModule,
+		deactivateModule,
+		regeneratePostByEmail,
+	}
```
… and same as here https://github.com/Automattic/wp-calypso/pull/90489/files#r1604574276, we should use `useDispatch` from `calypso/state` instead of `connect`s `mapDispatchToProps`

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,160 @@
+								return;
+							}
+
+							if ( checked ) {
+								activateModule( selectedSiteId, moduleSlug );
+							} else {
+								deactivateModule( selectedSiteId, moduleSlug );
+							}
```
I know the back-end for this isn't available yet, but conceptually, the `useQuery` and `useMutation` implementations in #90489 are great inspirations. Until the back-end is available, we can use mock implementations of those hooks.

nightnei committed:
refactor: address Fredriks feedback

nightnei committed:
chore: rename

chad1008 committed:
useSelect/useDispatch instead of connect

chad1008 commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,138 @@
+	} );
+
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+
+	const handleRegenerate = () => {
+		if ( ! selectedSiteId ) {
+			return;
+		}
+
+		const regenerate = regeneratePostByEmail( selectedSiteId );
+
+		dispatch( regenerate );
+	};
```
Real talk: I'm very new to redux, and this approach came from reading and some trial and error. The fact that it works feels like cheating at worst and a little too easy at best. Please don't hesitate to to tell me where this has gone amusingly sideways. My brain is clay for you to mold 😆

chad1008 commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,138 @@
+								return;
+							}
+
+							if ( checked ) {
+								activateModule( selectedSiteId, moduleSlug );
+							} else {
+								deactivateModule( selectedSiteId, moduleSlug );
+							}
```
Similar to my other comment about my redux-fu, the fact that this works feels like i've accidentally tapped into some dark magic. Previously these two methods were piped through `connect`, but now they just work and I don't know why. I hate it. Please send help.

nightnei replied:
Your code should not work, since `activateModule ` does nothing, it just returns another function.
To run it in context of redux you need to provide it to `dispatch`, like you did here (https://github.com/Automattic/wp-calypso/pull/90819#discussion_r1609061999)

`dispatch(activateModule( selectedSiteId, moduleSlug ))`

And it's the reason and the answer - why the case with `regeneratePostByEmail ` works well - there is `dispatch` wrapper function.
Hope it answers both your questions 🙂

chad1008 replied:
Thank you, that makes more sense, but also didn't at first because the toggle _did_ still work for JP/WoA sites. 

Then it dawned on me: this broken code was never being run. We don't actually need `activateModule` or `deactivateModule` because they're already built into `JetpackModuleToggle`. Putting them in my `onChange` handler didn't impact anything because that handler is only being run for simple sites. Mildly embarrassing, but hey... my handler is less verbose now. I'll call that a win.

chad1008 committed:
test new endpoints

nightnei commented on client/my-sites/site-settings/publishing-tools/index.jsx:
```diff
@@ -165,7 +100,7 @@ class PublishingTools extends Component {
 
 		const renderPressThis = config.isEnabled( 'press-this' ) && ! this.isMobile();
-		const renderPostByEmail = siteIsJetpack;
+		const renderPostByEmail = true; //TODO: previously `siteIsJetpack`, is any check needed going forward?
```
We can remove it

chad1008 replied:
Ah, thanks for confirming. I actually meant to reevaluate that and remove it (i was pretty sure it could go) before pushing the latest changes, but I forgot I'd left myself that note. 😆

chad1008 committed:
remove `siteIsJetpack` condition for post by email rendering

chad1008 committed:
remove jetpack activateModule/deactivateModule

chad1008 committed:
stop renaming variables like a cotton-headed ninny-muggins

chad1008 committed:
rename variables based on simple/jetpack path

chad1008 committed:
DRY condition for disabling DOM elements

chad1008 committed:
undo overzealous search and replace

chad1008 commented:
Thanks again for looking this over @nightnei. I've addressed your feedback and also gone through and cleaned things up a bit.

A big part of that was making sure the variables were named it ways that made it clear which path they were a part of (e.g. `jetpackRegeneratingPostByEmail` instead of simply `regeneratingPostByEmail`, because for simple sites, the regenerate button goes through a completely separate process with its own variables.

This means the variable names are more verbose than they were before. I _think_ the names I've chosen make the code easier to read, but please let me know if you disagree of if I've overcompensated!

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,183 @@
+const moduleSlug = 'post-by-email';
+
+export const PostByEmailSetting = ( { isFormPending, address }: PostByEmailSettingProps ) => {
+	const {
+		selectedSiteId,
+		siteIsJetpack,
+		siteIsAtomic,
+		jetpackRegeneratingPostByEmail,
+		jetpackPostByEmailIsActive,
+		moduleUnavailable,
+	} = useSelector( ( state ) => {
+		const selectedSiteId = getSelectedSiteId( state ) || 0;
+		const siteIsJetpack = isJetpackSite( state, selectedSiteId );
+		const siteIsAtomic = isSiteAutomatedTransfer( state, selectedSiteId );
+		const jetpackRegeneratingPostByEmail = isRegeneratingJetpackPostByEmail(
+			state,
+			selectedSiteId
+		);
+		const jetpackPostByEmailIsActive = !! isJetpackModuleActive(
+			state,
+			selectedSiteId,
+			moduleSlug
+		);
+		const siteInDevMode = isJetpackSiteInDevelopmentMode( state, selectedSiteId );
+		const moduleUnavailableInDevMode = isJetpackModuleUnavailableInDevelopmentMode(
+			state,
+			selectedSiteId,
+			moduleSlug
+		);
+
+		return {
+			selectedSiteId,
+			siteIsJetpack,
+			siteIsAtomic,
+			jetpackRegeneratingPostByEmail,
+			jetpackPostByEmailIsActive,
+			moduleUnavailable: siteInDevMode && moduleUnavailableInDevMode,
+		};
+	} );
```
```suggestion
	const selectedSiteId = useSelector( getSelectedSiteId ) || 0;
	const siteIsJetpack = useSelector( ( state ) => isJetpackSite( state, selectedSiteId ) );
	const siteIsAtomic = useSelector( ( state ) => isSiteAutomatedTransfer( state, selectedSiteId ) );
	const jetpackRegeneratingPostByEmail = useSelector( ( state ) =>
		isRegeneratingJetpackPostByEmail( state, selectedSiteId )
	);
	const jetpackPostByEmailIsActive = !! useSelector( ( state ) =>
		isJetpackModuleActive( state, selectedSiteId, moduleSlug )
	);
	const siteInDevMode = useSelector( ( state ) =>
		isJetpackSiteInDevelopmentMode( state, selectedSiteId )
	);
	const moduleUnavailableInDevMode = useSelector( ( state ) =>
		isJetpackModuleUnavailableInDevelopmentMode( state, selectedSiteId, moduleSlug )
	);
	const moduleUnavailable = siteInDevMode && moduleUnavailableInDevMode;
```
`useSelector` should be used for every piece of state we want to select.

chad1008 replied:
thanks for this. prompted me to spend a little more time with React Redux docs, which is always helpful.

chad1008 committed:
separate useSelector() calls

Co-authored-by: Fredrik Rombach Ekelund <fredrik@fredrik.computer>

nightnei commented:
Regarding test instructions - sometimes a reviewer is not in the context of the changes, honestly even for me it's totally not clear what is `wp-admin/admin.php?page=jetpack#/writing`, how to get to this page (direct link doesn't work for me) 😅

>Running Calypso locally off of this branch, visit Settings > Writing for a simple site, and make sure you have the Classic View selected (Calypso, not wp-admin)

Here is typo, right? ~Classic~ Default


UPDATE:
I have tested and I still don't understand what is `wp-admin/admin.php?page=jetpack#/writing` 😅

How I would rewrite your test instructions here (it's just a draft):

>Find your simple test site in the dashboard and note the state and/or email address of the current post by email settings
Back in Calypso, test the Post by Email toggle, refreshing and checking dashboard.wordpress.com to confirm the settings from Calypso are carrying over.
Back in Calypso, test the Regenerate button. Confirm that a new address is generated, and that the new address if reflected when refreshing dashboard.wordpress.com

### Test steps for Simple website
1) Open Setting -> Writting in Default (Calypso )
2) Open the second tab `https://dashboard.wordpress.com/` and find your website
3) Now we will be asserting that all changes applied in Calypso are synced with dashboard:
3.1) Turn on "Post by Email", assert that email is generated and the same email is rendered in dashboard
3.2) Regenerate email and assert ...
3.3) Turn off "Post by Email" and assert ...
4) Now let's repeat 3.x steps but making changes to dashboard and assert that they are reflected to Calypso

### Test steps for Atomic website
>1)Switch over to Settings > Writing for a WoA site
2) Repeat the same tests, this time comparing them to wp-admin/admin.php?page=jetpack#/writing, making sure the Calypso settings still properly update the Jetpack settings (we didn't make changes here, but the code was moved and refactored so we're making sure nothing got broken in the process)

Offtopic: Newbie question - WoA is WordPress on Atomic, right?

Here I would specify that "now we will be testing Atomic website, so necessary to create another website and make it Atomic", since initially I was confused with short sentense and so quick and dramatic "jump" from Simple to Atomic, w/o explanation that now we will be testing another thing.
And I still don't know what is `wp-admin/admin.php?page=jetpack#/writing` 😅

nightnei commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,163 @@
+	);
+	const moduleUnavailable = siteInDevMode && moduleUnavailableInDevMode;
+	const { data: simpleSitePostByEmailSettings } = useGetPostByEmail( selectedSiteId );
+	const { mutate: simpleSiteSwitchPostByEmail, isPending: isSimpleSitePendingSwitch } =
```
replace everywhere `switch` with `toggle` for consistency

chad1008 replied:
Thanks! I actually went with `switch` specifically to be consistent with your PR, and didn't see the update to `toggle` until I did my rebase just now.

nightnei commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,163 @@
+		} );
+	};
+
+	const email = address && address !== 'regenerate' ? address : '';
```
Below we have extra condition and additionallt this variable doesn't make sense to generate for `! siteIsJetpack` website. So let's keep 1 condition and generate `email` in one place;
```suggestion
	let email;
	if (siteIsJetpack) {
		email = address && address !== 'regenerate' ? address : '';
	} else {
		email = simpleSitePostByEmailSettings?.email;
	}
```

nightnei commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,163 @@
+					<ClipboardButtonInput
+						className="publishing-tools__email-address"
+						disabled={ jetpackRegeneratingPostByEmail || ! isActive || moduleUnavailable }
+						value={ siteIsJetpack ? email : simpleSitePostByEmailSettings?.email }
```
```suggestion
						value={ email }
```

nightnei commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,163 @@
+	const isActive = siteIsJetpack
+		? jetpackPostByEmailIsActive
+		: simpleSitePostByEmailSettings?.isEnabled;
+	const labelClassName =
```
Let's use `clsx` instead of manually handling class name

fredrikekelund replied:
Since `clsx` hasn't landed yet (waiting on #89843), I made a change in 09e5011ec0 to use `classNames` as the import name instead.

nightnei commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,163 @@
+					/>
+					<Button
+						onClick={ handleRegenerate }
+						disabled={
+							isFormPending ||
+							jetpackRegeneratingPostByEmail ||
+							isSimpleSitePendingRegenerate ||
+							! isActive ||
+							!! moduleUnavailable
```
We have different condition for disabled state here and here (https://github.com/Automattic/wp-calypso/pull/90819/files#diff-1a2d2a711ac2389148e7764b72177d565e3e6a88fea709f208906a031a5a64c5R142)

1) Let's make it the same (I thing we can add `isFormPending` to the "Copy" button)
2) Let's create variable for both cases to avoid duplication

chad1008 replied:
This one was actually a bit of an interesting rabbit hole once I started looking at it more closely.

There are five `disabled` props in play:
1. Jetpack toggle
2. Simple site toggle
3. Copy button,
4. Regenerate button
5. `is-disabled` classname on the `FormLabel`. 
6. 
7. Across those five items, we actually had _four_ different conditions. Technically only four ever appear at once, but still... that's way too many combinations of conditions!

To clean this up, I've made it so the three items that apply to both flows are the same using a new, single variable to check the relevant conditions. I've also made it so `isFormPending` is included in all conditions, as that's not flow-specific and checks to see if any settings are being requested for the current site.

One other interesting discovery... for the new Simple site toggle, we disable the toggle itself if the setting is being toggled or if the email is being regenerated. I think that's something I initially included when taking inspiration from earlier versions of your Post by Voice PR.

We don't actually have the same behavior for Jetpack, because:
1. We don't have a Jetpack version of `isSimpleSitePendingToggle` (the `isUpdatingJetpackSettings` selector might suffice, but I'll need to look at it a bit more)
2. While we do have `jetPackRegeneratingPostByEmail`, I don't see it actually _doing_ anything. It doesn't apply a disabled state, or change the button text the way the code currently asks it to. It doesn't even log out as `true` when the regeneration button is pressed. I'm not sure if it just isn't implemented correctly, or if there's some other problem. 

When I'm back on tuesday, I'll prioritize looking into `jetPackRegeneratingPostByEmail` and `isUpdatingJetpackSettings` to see if they can serve the same roles as our Simple counterparts, or if they should just be excluded entirely for now.

But either way, the disabled logic should be a little simpler now... and might become more so if we decide to go without `jetPackRegeneratingPostByEmail`.

nightnei reviewed (changes requested):
In general everything look cool 👍 And works will (both UI and backend). I haven't only tested the last two steps in your tet instructions, but I tested PbE in Default(Calypso) View and it works well, as previously.
Left a few comments and will be waiting for rebasing with `trunk` to make a final review.

chad1008 committed:
Vova's feedback: clsx and variable renaming

chad1008 committed:
clean up disabled state conditions

chad1008 committed:
from feedback: simplify `email` declaration

Co-authored-by: Volodymyr Makukha <nei.css@gmail.com>

chad1008 commented:
Thank you @nightnei! I appreciate the feedback on the testing instructions, and have made some changes there 🙂

> Offtopic: Newbie question - WoA is WordPress on Atomic, right?

Yep! Sorry for throwing the extra acronym around!

-- 

Thanks for all of the other feedback, I believe I've addressed everything, I'll just need to circle back to the `disabled` logic question I mentioned above. Will do that first thing when I'm back online!

chad1008 committed:
fix lint error from feedback commit

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/index.jsx:
```diff
@@ -178,7 +112,7 @@ class PublishingTools extends Component {
 				<SettingsSectionHeader title={ translate( 'Publishing Tools' ) } />
 
 				{ renderPostByVoice && this.renderPostByVoiceModule() }
-				{ renderPostByEmail && this.renderPostByEmailModule() }
+				{ this.renderPostByEmailModule() }
```
I think we should put Post by Email above Post by Voice since it's used much more frequently.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,168 @@
+	const moduleUnavailableInDevMode = useSelector( ( state ) =>
+		isJetpackModuleUnavailableInDevelopmentMode( state, selectedSiteId, moduleSlug )
+	);
+	const moduleUnavailable = siteInDevMode && moduleUnavailableInDevMode;
+	const { data: simpleSitePostByEmailSettings } = useGetPostByEmail( selectedSiteId );
```
```suggestion
	const moduleUnavailable = siteInDevMode && moduleUnavailableInDevMode;

	const { data: simpleSitePostByEmailSettings } = useGetPostByEmail( selectedSiteId );
```
Nit, but a newline here would be nice to delineate between Jetpack and Simple stuff.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,168 @@
+					<ToggleControl
+						checked={ !! simpleSitePostByEmailSettings?.isEnabled }
+						disabled={ isFormPending || isSimpleSitePendingToggle || isSimpleSitePendingRegenerate }
+						label={ translate( 'Post by Email' ) }
```
```suggestion
						label={ translate( 'Publish posts by sending an email' ) }
```
It seems logical to reuse the Jetpack control label here.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,168 @@
+	const selectedSiteId = useSelector( getSelectedSiteId ) || 0;
+	const siteIsJetpack = useSelector( ( state ) => isJetpackSite( state, selectedSiteId ) );
+	const siteIsAtomic = useSelector( ( state ) => isSiteAutomatedTransfer( state, selectedSiteId ) );
+	const jetpackRegeneratingPostByEmail = useSelector( ( state ) =>
+		isRegeneratingJetpackPostByEmail( state, selectedSiteId )
+	);
```
This is outside the scope of this PR, but in my testing, this selector doesn't seem to work properly (it never returns `true`). The same behavior can be observed in production.

chad1008 replied:
Do you think it's worth removing from Post by Email for now, so it's not misleading/confusing for devs reading this code in the future? Or leave it to quietly do nothing in case/until it's fixed in the future?

chad1008 replied:
Scratch that question... we were really [only trying to use it for one thing that I decided to remove anyway](https://github.com/Automattic/wp-calypso/pull/90819/files#r1617782153), so that selector is no longer called at all. If it's ever fixed we can bring it back to have parity with Simple sites, but for now we can leave it out, I think.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,168 @@
+	const isActive = siteIsJetpack
+		? jetpackPostByEmailIsActive
+		: simpleSitePostByEmailSettings?.isEnabled;
+	const isDisabledForBothFlows =
+		isFormPending ||
+		jetpackRegeneratingPostByEmail ||
+		isSimpleSitePendingRegenerate ||
+		! isActive ||
+		!! moduleUnavailable;
+
+	return (
+		<>
+			<FormFieldset>
+				<SupportInfo
+					text={ translate(
+						'Allows you to publish new posts by sending an email to a special address.'
+					) }
+					link={
+						siteIsAtomic
+							? localizeUrl( 'https://wordpress.com/support/post-by-email/' )
+							: 'https://jetpack.com/support/post-by-email/'
+					}
+					privacyLink={ ! siteIsAtomic }
+				/>
+				{ siteIsJetpack ? (
+					<JetpackModuleToggle
+						siteId={ selectedSiteId }
+						moduleSlug="post-by-email"
+						label={ translate( 'Publish posts by sending an email' ) }
+						disabled={ isFormPending || moduleUnavailable }
+					/>
+				) : (
+					<ToggleControl
+						checked={ !! simpleSitePostByEmailSettings?.isEnabled }
+						disabled={ isFormPending || isSimpleSitePendingToggle || isSimpleSitePendingRegenerate }
```
```suggestion
	const isJetpackDisabled = isFormPending || moduleUnavailable;
	const isSimpleSiteDisabled =
		isFormPending || isSimpleSitePendingToggle || isSimpleSitePendingRegenerate;

	const isDisabledControls = ! isActive || isJetpackDisabled || isSimpleSiteDisabled;

	return (
		<>
			<FormFieldset>
				<SupportInfo
					text={ translate(
						'Allows you to publish new posts by sending an email to a special address.'
					) }
					link={
						siteIsAtomic
							? localizeUrl( 'https://wordpress.com/support/post-by-email/' )
							: 'https://jetpack.com/support/post-by-email/'
					}
					privacyLink={ ! siteIsAtomic }
				/>

				{ siteIsJetpack ? (
					<JetpackModuleToggle
						siteId={ selectedSiteId }
						moduleSlug="post-by-email"
						label={ translate( 'Publish posts by sending an email' ) }
						disabled={ isJetpackDisabled }
					/>
				) : (
					<ToggleControl
						checked={ !! simpleSitePostByEmailSettings?.isEnabled }
						disabled={ isSimpleSiteDisabled }
```
It'd be nice to assign the `disabled` values for the respective controls to separate variables.

chad1008 replied:
Oh yes, that's much cleaner, thank you!

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/hooks/use-get-post-by-email.tsx:
```diff
@@ -0,0 +1,34 @@
+
+export const getPostByEmailPath = ( siteId: number | null ) => `/sites/${ siteId }/post-by-email`;
+
+export const getCachePostByEmailKey = ( siteId: number | null ) => [
```
```suggestion
export const getPostByEmailKeyQueryKey = ( siteId: number | null ) => [
```
I overlooked this in #90489, but I think we should mention "query key" rather than "cache key".

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/index.jsx:
```diff
@@ -21,6 +15,9 @@ import isRegeneratingJetpackPostByEmail from 'calypso/state/selectors/is-regener
 import isSiteAutomatedTransfer from 'calypso/state/selectors/is-site-automated-transfer';
 import { isJetpackSite } from 'calypso/state/sites/selectors';
 import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+import PressThis from '../press-this';
+import { PostByEmailSetting } from './post-by-email';
+import { PostByVoiceSetting } from './post-by-voice';
```
```suggestion
import PressThis from 'calypso/my-sites/site-settings/press-this';
import { PostByEmailSetting } from 'calypso/my-sites/site-settings/publishing-tools/post-by-email';
import { PostByVoiceSetting } from 'calypso/my-sites/site-settings/publishing-tools/post-by-voice';
```
Nit, but let's be consistent with absolute import paths.

chad1008 replied:
Shoot! I made a point of fixing that, because it was mentioned on a previous PR. I must have lost it during a rebase 🤦

fredrikekelund reviewed (commented):
Looking good and tests well! 👍

I left a couple of small remarks.

chad1008 committed:
address Fredrik feedback

chad1008 committed:
clean up disabled props, only better

chad1008 committed:
remove `isRegeneratingJetpackPostByEmail` and related logic

chad1008 commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,162 @@
+						value={ email }
+					/>
+					<Button onClick={ handleRegenerate } disabled={ isDisabledControls }>
+						Regenerate address
```
There was previously (including on `trunk`) logic to change this button content to `Regenerating...` while a new address was being generated for Jetpack sites. I've removed that in favor of consistent button text for the reasons below.

It isn't working for Jetpack (including on trunk) as Fredrik and I noted above, and when I applied it to Simple sites (where it _does_ work) I really didn't like the way it changed the size of the button. The visual jump was king of jarring. Could be fixed with CSS, but it honestly didn't feel like it was worth it.

On Simple, we're also now temporarily disabling the button while the new address is generated, so there's visual feedback in that form, and if/when the `isRegeneratingJetpackPostByEmail` is working in the future, we can apply the temp disable effect there, too.

tl;dr: from a user perspective, removing this text change doesn't impact Jetpack users at all, and we have something else in place for Simple that I think works better anyway.

chad1008 commented:
Current feedback has been addressed, and I've added [one new inline comment](https://github.com/Automattic/wp-calypso/pull/90819/files#r1617782153) regarding a small change when regenerating the email address.

Based on past feedback I think we should be good, or very very close!

fredrikekelund committed:
Small tweaks

fredrikekelund committed:
Simplify PublishingTools component

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/index.jsx:
```diff
@@ -165,66 +53,30 @@ class PublishingTools extends Component {
+	siteIsJetpack: PropTypes.bool,
 };
 
-export default connect(
-	( state ) => {
-		const selectedSiteId = getSelectedSiteId( state );
-		const siteIsJetpack = isJetpackSite( state, selectedSiteId );
-		const isAtomic = isSiteAutomatedTransfer( state, selectedSiteId );
-		const regeneratingPostByEmail = isRegeneratingJetpackPostByEmail( state, selectedSiteId );
-		const siteInDevMode = isJetpackSiteInDevelopmentMode( state, selectedSiteId );
-		const moduleUnavailableInDevMode = isJetpackModuleUnavailableInDevelopmentMode(
-			state,
-			selectedSiteId,
-			'post-by-email'
-		);
-
-		return {
-			siteIsJetpack,
-			isAtomic,
-			selectedSiteId,
-			regeneratingPostByEmail,
-			postByEmailAddressModuleActive: !! isJetpackModuleActive(
-				state,
-				selectedSiteId,
-				'post-by-email'
-			),
-			moduleUnavailable: siteInDevMode && moduleUnavailableInDevMode,
-		};
-	},
-	{
-		regeneratePostByEmail,
-	}
-)( localize( PublishingTools ) );
```
My goal here wasn't to remove the `connect` call. However, I discovered that several props were redundant (a lot of logic has been moved to `PostByEmailSetting`) and that `SiteSettingsFormWriting` can pass the other props we need.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,162 @@
+export const PostByEmailSetting = ( { emailAddress }: PostByEmailSettingProps ) => {
+	const selectedSiteId = useSelector( getSelectedSiteId ) || 0;
+	const siteIsJetpack = useSelector( ( state ) => isJetpackSite( state, selectedSiteId ) );
+	const jetpackRegeneratingPostByEmail = siteIsJetpack && emailAddress === 'regenerate';
```
The `isRegeneratingJetpackPostByEmail` selector doesn't work, but we can deduce the same thing by determining if `emailAddress` has the value `regenerate`. "One weird trick" 🙃

chad1008 replied:
This almost feels sneaky, and I love it 😆

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/post-by-email.tsx:
```diff
@@ -0,0 +1,162 @@
+					text={ translate(
+						'Allows you to publish new posts by sending an email to a special address.'
+					) }
+					link={
+						! siteIsJetpack
+							? localizeUrl( 'https://wordpress.com/support/post-by-email/' )
+							: 'https://jetpack.com/support/post-by-email/'
+					}
+					privacyLink={ siteIsJetpack }
```
We used to look at `siteIsAtomic` here and the logic was flipped (Simple sites linked to jetpack.com and Atomic sites to wordpress.com).

The flipped logic was obviously wrong, but I also don't see a reason to look at `siteIsAtomic` instead of `siteIsJetpack`.

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/index.jsx:
```diff
@@ -29,113 +17,13 @@ class PublishingTools extends Component {
 		return /Mobile|iP(hone|od|ad)|Android|BlackBerry|IEMobile|Silk/.test( navigator.userAgent );
 	}
 
-	componentDidUpdate() {
-		const {
-			fields,
-			moduleUnavailable,
-			postByEmailAddressModuleActive,
-			regeneratingPostByEmail,
-			selectedSiteId,
-		} = this.props;
-
-		if ( ! moduleUnavailable ) {
-			return;
-		}
-
-		if (
-			postByEmailAddressModuleActive &&
-			regeneratingPostByEmail === null &&
-			! fields.post_by_email_address
-		) {
-			this.props.regeneratePostByEmail( selectedSiteId );
-		}
-	}
-
-	onRegenerateButtonClick = () => {
-		this.props.regeneratePostByEmail( this.props.selectedSiteId );
-	};
-
-	isFormPending() {
-		const { isRequestingSettings, isSavingSettings } = this.props;
-
-		return isRequestingSettings || isSavingSettings;
-	}
```
Initially, I thought we could remove this altogether, but I realized that the `componentDidUpdate` logic is still required. What it does is regenerate the email address immediately after the Jetpack Post by Email module is enabled for the first time. Without it, the feature can be enabled, but there's no email address to send to.

I moved this functionality to `PostByEmailSetting` in 2c7e20c3d2349a207a6f10fcd0be8fe83bf9ad48

fredrikekelund commented on client/my-sites/site-settings/publishing-tools/hooks/use-get-post-by-voice.tsx:
```diff
@@ -8,15 +8,15 @@ import type {
 export const getPostByVoicePath = ( siteId: number | null ) => `/sites/${ siteId }/post-by-voice`;
 
-export const getCachePostByVoiceKey = ( siteId: number | null ) => [
+export const getPostByVoiceQueryKey = ( siteId: number | null ) => [
```
For consistency, I took the opportunity to rename this.

fredrikekelund committed:
Fix regression

fredrikekelund committed:
Translation comments

fredrikekelund reviewed (approved):
I made some light tweaks in 09e5011ec079f59612f40781ecabf652e52f37cd and then some slightly larger refactorings of the `PublishingTools` component in 6cf91ac43ee6bb65a15f177d2cf06a1fa94db8b2 and 2c7e20c3d2349a207a6f10fcd0be8fe83bf9ad48. All of the Post by Email-related logic is consolidated in `PostByEmailSetting` now.

I've tested this with a Simple site, an Atomic site, and an external Jetpack site.

This should be good to land 👍
